### ISSUE TITLE
Indexes 'rust_expanded_macros' forever

### ISSUE THREAD

**Issue by ralfbiedert** [intellij-rust/intellij-rust#8020]
## Environment

* **IntelliJ Rust plugin version:** 0.4.158.4203-212
* **Rust toolchain version:** 1.58.0-nightly (00d5e42e7 2021-10-24) x86_64-pc-windows-msvc
* **IDE name and version:** CLion 2021.2.3 (CL-212.5457.51)
* **Operating system:** Windows 10 10.0
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description

Coincidentally while writing my last ticket I got a power failure. After restarting CLion I ended up in a state where keeps trying to "Update indexes".

![image](https://user-images.githubusercontent.com/301288/139236900-3b411488-ff5f-4865-b008-726b3dc74859.png)

![image](https://user-images.githubusercontent.com/301288/139236803-5217bb19-d23b-45d5-a18e-daed6679e7a8.png)

This pretty much goes on forever and iterates about 20 files under `\rust_expanded_macros\...`. Deleting my `target` folder didn't help.

Update, deleting my `target` and `.idea` folder seems to have fixed it.

**Comment by ralfbiedert** [comment]
When investigating #8149 it got stuck on "Analyzing" again. I happened to have the "Tail" window for the console open and noticed these ones here got spammed all the time (like 2 per second):

```
2021-11-22 20:07:41,066 [ 206951]  ERROR - stubs.StubProcessingHelperBase - CLion 2021.2.3  Build #CL-212.5457.51 
2021-11-22 20:07:41,066 [ 206951]  ERROR - stubs.StubProcessingHelperBase - JDK: 11.0.12; VM: OpenJDK 64-Bit Server VM; Vendor: JetBrains s.r.o. 
2021-11-22 20:07:41,066 [ 206951]  ERROR - stubs.StubProcessingHelperBase - OS: Windows 10 
2021-11-22 20:07:41,066 [ 206951]  ERROR - stubs.StubProcessingHelperBase - Plugin to blame: Rust version: 0.4.160.4261-212 
2021-11-22 20:07:41,066 [ 206951]  ERROR - stubs.StubProcessingHelperBase - Last Action: TailIntelliJLog 
2021-11-22 20:07:41,066 [ 206951]  ERROR - stubs.StubProcessingHelperBase - Stub index points to a file (file = file://C:/Users/rb/.rustup/toolchains/nightly-x86_64-pc-windows-msvc/lib/rustlib/src/rust/library/std/src/os/freebsd/mod.rs, file type = org.rust.lang.RsFileType@7d6e6153, indexed file type = org.rust.lang.RsFileType@7d6e6153) without indexed stub tree; indexing stamp = indexing timestamp = 1637583816840, binary = false, byte size = 109, char size = 109, can have stubs = true, actual stub count = 0 
java.lang.Throwable: Stub index points to a file (file = file://C:/Users/rb/.rustup/toolchains/nightly-x86_64-pc-windows-msvc/lib/rustlib/src/rust/library/std/src/os/freebsd/mod.rs, file type = org.rust.lang.RsFileType@7d6e6153, indexed file type = org.rust.lang.RsFileType@7d6e6153) without indexed stub tree; indexing stamp = indexing timestamp = 1637583816840, binary = false, byte size = 109, char size = 109, can have stubs = true, actual stub count = 0
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:182)
	at com.intellij.psi.stubs.StubProcessingHelper.retrieveStubIdList(StubProcessingHelper.java:42)
	at com.intellij.psi.stubs.StubIndexImpl.lambda$processElements$4(StubIndexImpl.java:326)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
	at com.intellij.psi.stubs.StubIndexImpl.processElements(StubIndexImpl.java:325)
	at com.intellij.psi.stubs.StubIndex.processElements(StubIndex.java:51)
	at org.rust.lang.core.stubs.index.RsModulesIndex$Companion.getDeclarationsFor$processElements(RsModulesIndex.kt:43)
	at org.rust.lang.core.stubs.index.RsModulesIndex$Companion.getDeclarationsFor(RsModulesIndex.kt:52)
	at org.rust.lang.core.psi.RsFile._get_declarations_$lambda-4(RsFile.kt:248)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:114)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:43)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:68)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:228)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at org.rust.lang.core.psi.RsFile.getDeclarations(RsFile.kt:247)
	at org.rust.lang.core.psi.RsFile.getDeclaration(RsFile.kt:231)
	at org.rust.lang.core.psi.RsFile.doGetCachedData(RsFile.kt:115)
	at org.rust.lang.core.psi.RsFile._get_cachedData_$lambda-1$lambda-0(RsFile.kt:93)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:114)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:43)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:68)
	at org.rust.openapiext.UtilsKt.recursionGuard(utils.kt:90)
	at org.rust.openapiext.UtilsKt.recursionGuard$default(utils.kt:89)
	at org.rust.lang.core.psi.RsFile._get_cachedData_$lambda-1(RsFile.kt:93)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:114)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:43)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:68)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:228)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at org.rust.lang.core.psi.RsFile.getCachedData(RsFile.kt:92)
	at org.rust.lang.core.psi.RsFile.getCrate(RsFile.kt:76)
	at org.rust.lang.core.psi.ext.RsElementKt.getContainingCrate(RsElement.kt:58)
	at org.rust.lang.core.stubs.common.RsAttributeOwnerPsiOrStub.getContainingCrate(PsiOrStubInterfaces.kt:49)
	at org.rust.lang.core.psi.ext.RsDocAndAttributeOwnerKt.evaluateCfg(RsDocAndAttributeOwner.kt:475)
	at org.rust.lang.core.psi.ext.CfgUtilsKt.getCodeStatus(CfgUtils.kt:54)
	at org.rust.lang.core.psi.ext.CfgUtilsKt.getExistsAfterExpansion(CfgUtils.kt:30)
	at org.rust.lang.core.psi.RsFileKt.isValidProjectMember(RsFile.kt:311)
	at org.rust.lang.core.resolve.RsCachedTypeAlias$isFreeAndValid$2.invoke(RsCachedTypeAlias.kt:33)
	at org.rust.lang.core.resolve.RsCachedTypeAlias$isFreeAndValid$2.invoke(RsCachedTypeAlias.kt:30)
	at kotlin.SafePublicationLazyImpl.getValue(LazyJVM.kt:107)
	at org.rust.lang.core.resolve.RsCachedTypeAlias.isFreeAndValid(RsCachedTypeAlias.kt:30)
	at org.rust.lang.core.resolve.indexes.RsTypeAliasIndex$Companion.findPotentialAliases(RsTypeAliasIndex.kt:41)
	at org.rust.lang.core.resolve.ImplLookup.processTyFingerprintsWithAliases(ImplLookup.kt:510)
	at org.rust.lang.core.resolve.ImplLookup.findExplicitImpls(ImplLookup.kt:486)
	at org.rust.lang.core.resolve.ImplLookup.rawFindImplsAndTraits(ImplLookup.kt:343)
	at org.rust.lang.core.resolve.ImplLookup.access$rawFindImplsAndTraits(ImplLookup.kt:247)
	at org.rust.lang.core.resolve.ImplLookup$findImplsAndTraits$cached$1.invoke(ImplLookup.kt:308)
	at org.rust.lang.core.resolve.ImplLookup$findImplsAndTraits$cached$1.invoke(ImplLookup.kt:308)
	at org.rust.stdext.Cache$Companion$fromConcurrentMap$1.getOrPut(Cache.kt:37)
	at org.rust.lang.core.resolve.ImplLookup.findImplsAndTraits(ImplLookup.kt:308)
	at org.rust.lang.core.resolve.NameResolutionKt.processAssociatedItems(NameResolution.kt:1376)
	at org.rust.lang.core.resolve.NameResolutionKt.processMethodDeclarationsWithDeref(NameResolution.kt:1348)
	at org.rust.lang.core.resolve.NameResolutionKt.processMethodCallExprResolveVariants(NameResolution.kt:154)
	at org.rust.lang.core.resolve.ref.RsMethodCallReferenceImplKt$resolveMethodCallReferenceWithReceiverType$1.invoke(RsMethodCallReferenceImpl.kt:54)
	at org.rust.lang.core.resolve.ref.RsMethodCallReferenceImplKt$resolveMethodCallReferenceWithReceiverType$1.invoke(RsMethodCallReferenceImpl.kt:53)
	at org.rust.lang.core.resolve.ProcessorsKt.collectResolveVariantsAsScopeEntries(Processors.kt:155)
	at org.rust.lang.core.resolve.ref.RsMethodCallReferenceImplKt.resolveMethodCallReferenceWithReceiverType(RsMethodCallReferenceImpl.kt:53)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferMethodCallExprType(TypeInferenceWalker.kt:623)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferDotExprType(TypeInferenceWalker.kt:828)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:173)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferTypes(TypeInferenceWalker.kt:949)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferIfExprType(TypeInferenceWalker.kt:927)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:176)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType$default(TypeInferenceWalker.kt:155)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.processStatement(TypeInferenceWalker.kt:151)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferType(TypeInferenceWalker.kt:108)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferTypeCoercableTo(TypeInferenceWalker.kt:101)
	at org.rust.lang.core.types.infer.RsTypeInferenceWalker.inferFnBody(TypeInferenceWalker.kt:96)
	at org.rust.lang.core.types.infer.RsInferenceContext.infer(TypeInference.kt:174)
	at org.rust.lang.core.types.infer.TypeInferenceKt.inferTypesIn$lambda-0(TypeInference.kt:33)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:114)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:43)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:68)
	at org.rust.openapiext.UtilsKt.recursionGuard(utils.kt:90)
	at org.rust.lang.core.types.infer.TypeInferenceKt.inferTypesIn(TypeInference.kt:33)
	at org.rust.lang.core.types.ExtensionsKt._get_inference_$lambda-0(Extensions.kt:75)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:114)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:43)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:68)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:228)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at org.rust.lang.core.types.ExtensionsKt.getInference(Extensions.kt:74)
	at org.rust.lang.core.types.ExtensionsKt.getInference(Extensions.kt:81)
	at org.rust.lang.core.types.ExtensionsKt.getType(Extensions.kt:93)
	at org.rust.lang.core.dfa.CFGBuilder.processCall(CFGBuilder.kt:193)
	at org.rust.lang.core.dfa.CFGBuilder.visitDotExpr(CFGBuilder.kt:582)
	at org.rust.lang.core.psi.impl.RsDotExprImpl.accept(RsDotExprImpl.java:27)
	at org.rust.lang.core.psi.impl.RsDotExprImpl.accept(RsDotExprImpl.java:32)
	at org.rust.lang.core.dfa.CFGBuilder.process(CFGBuilder.kt:168)
	at org.rust.lang.core.dfa.CFGBuilder.visitIfExpr(CFGBuilder.kt:389)
	at org.rust.lang.core.psi.impl.RsIfExprImpl.accept(RsIfExprImpl.java:27)
	at org.rust.lang.core.psi.impl.RsIfExprImpl.accept(RsIfExprImpl.java:32)
	at org.rust.lang.core.dfa.CFGBuilder.process(CFGBuilder.kt:168)
	at org.rust.lang.core.dfa.CFGBuilder.visitExprStmt(CFGBuilder.kt:225)
	at org.rust.lang.core.psi.impl.RsExprStmtImpl.accept(RsExprStmtImpl.java:21)
	at org.rust.lang.core.psi.impl.RsExprStmtImpl.accept(RsExprStmtImpl.java:26)
	at org.rust.lang.core.dfa.CFGBuilder.process(CFGBuilder.kt:168)
	at org.rust.lang.core.dfa.CFGBuilder.visitBlock(CFGBuilder.kt:203)
	at org.rust.lang.core.psi.impl.RsBlockImpl.accept(RsBlockImpl.java:27)
	at org.rust.lang.core.psi.impl.RsBlockImpl.accept(RsBlockImpl.java:32)
	at org.rust.lang.core.dfa.CFGBuilder.process(CFGBuilder.kt:168)
	at org.rust.lang.core.dfa.ControlFlowGraph$Companion.buildFor(ControlFlowGraph.kt:87)
	at org.rust.lang.core.types.ExtensionsKt._get_controlFlowGraph_$lambda-3(Extensions.kt:151)
	at com.intellij.psi.util.CachedValuesManager$1.compute(CachedValuesManager.java:158)
	at com.intellij.psi.impl.PsiCachedValueImpl.doCompute(PsiCachedValueImpl.java:39)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$3(CachedValueBase.java:227)
	at com.intellij.util.CachedValueBase.computeData(CachedValueBase.java:42)
	at com.intellij.util.CachedValueBase.lambda$getValueWithLock$4(CachedValueBase.java:227)
	at com.intellij.openapi.util.RecursionManager$1.computePreventingRecursion(RecursionManager.java:114)
	at com.intellij.openapi.util.RecursionGuard.doPreventingRecursion(RecursionGuard.java:43)
	at com.intellij.openapi.util.RecursionManager.doPreventingRecursion(RecursionManager.java:68)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:228)
	at com.intellij.psi.impl.PsiCachedValueImpl.getValue(PsiCachedValueImpl.java:28)
	at com.intellij.util.CachedValuesManagerImpl.getCachedValue(CachedValuesManagerImpl.java:72)
	at com.intellij.psi.util.CachedValuesManager.getCachedValue(CachedValuesManager.java:155)
	at org.rust.lang.core.types.ExtensionsKt.getControlFlowGraph(Extensions.kt:149)
	at org.rust.ide.inspections.lints.RsUnreachableCodeInspection$buildVisitor$1.visitFunction(RsUnreachableCodeInspection.kt:32)
	at org.rust.lang.core.psi.impl.RsFunctionImpl.accept(RsFunctionImpl.java:27)
	at org.rust.lang.core.psi.impl.RsFunctionImpl.accept(RsFunctionImpl.java:32)
	at com.intellij.codeInspection.InspectionEngine.acceptElements(InspectionEngine.java:64)
	at com.intellij.codeInspection.InspectionEngine.createVisitorAndAcceptElements(InspectionEngine.java:55)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.runToolOnElements(LocalInspectionsPass.java:326)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$visitPriorityElementsAndInit$4(LocalInspectionsPass.java:285)
	at com.intellij.util.AstLoadingFilter.forceAllowTreeLoading(AstLoadingFilter.java:159)
	at com.intellij.util.AstLoadingFilter.forceAllowTreeLoading(AstLoadingFilter.java:151)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$visitPriorityElementsAndInit$5(LocalInspectionsPass.java:283)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:130)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:119)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$visitPriorityElementsAndInit$6(LocalInspectionsPass.java:283)
	at com.intellij.concurrency.ApplierCompleter.execAndForkSubTasks(ApplierCompleter.java:136)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1078)
	at com.intellij.concurrency.ApplierCompleter.lambda$wrapInReadActionAndIndicator$1(ApplierCompleter.java:92)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:705)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:647)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:63)
	at com.intellij.concurrency.ApplierCompleter.wrapInReadActionAndIndicator(ApplierCompleter.java:104)
	at com.intellij.concurrency.ApplierCompleter.lambda$compute$0(ApplierCompleter.java:83)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:174)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:183)
	at com.intellij.concurrency.ApplierCompleter.compute(ApplierCompleter.java:83)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

**Comment by ralfbiedert** [comment]
A temporary fix seems to be running these here:

![image](https://user-images.githubusercontent.com/301288/142921340-2263e40d-06bb-4b4f-bb54-75a378e031c6.png)

Once CLion restarts it takes ~2 minutes to index everything but without errors and afterwards everything seems to work.

**Comment by ralfbiedert** [comment]
This seems to be fixed with 2021.3 / 161.
